### PR TITLE
Workaround crash in msvc when raising a GEOS exception

### PR DESCRIPTION
**Pull request by nyalldawson**
In some circumstances MSVC tries to raise it's own internal exception when we try to raise a GEOSException. This results in a hard crash of QGIS.

To workaround this, we cautiously try to raise a GEOSException on msvc builds, catching all other raised exceptions if throwing the GEOSException triggers them. This at least allows us to gracefully handle the geos error without crashing QGIS (although the detailed GEOS error message is not available when this situation occurs... but that's still better than a crash!).

If you want to try to find a better fix for this, the testqgsexpression.cpp test suite will demonstrate the issue.

Without this fix the test will crash on the "line_interpolate_point point" test.

Refs https://issues.qgis.org/issues/14752

On a scale from 0 (not a hack) to 10 (horrible hack), this is a 12

### CODE REVIEW

File: src/core/geometry/qgsgeos.cpp [hunk 1]
```diff
@@ -58,10 +58,28 @@ static void throwGEOSException( const char *fmt, ... )
...
+  // see https://issues.qgis.org/issues/14752
+  // if you want to test alternative fixes for this, run the testqgsexpression.cpp test suite - that will crash
+  // and burn on the "line_interpolate_point point" test if a GEOSException is thrown.
+  try
+  {
+    throw GEOSException( message );
+  }
+  catch (...)
+  {
+    // oops, msvc threw an exception when we tried to throw the exception!
```
Comment (luipir) [inline src/core/geometry/qgsgeos.cpp]: Hi @nyalldawson I'm always impressed how you can find these fixes, congratulations... btw would make sense to add a TODO label in the code parts that have need more love. More love means: find the best fit to manage strange behaviour of MSVC.

Comment (nyalldawson) [reply src/core/geometry/qgsgeos.cpp]: I wouldn't call this "fix" impressive by any means, but thanks anyway! I'll add the to-do :)

Comment (luipir) [reply src/core/geometry/qgsgeos.cpp]: I'm impressed how to have a debug session to find this kinds of errors :)

**Review by m-kuhn — APPROVED** [review]
Awesome fix, can you upstream this to MSVC, so we can drop this band aid for QGIS 4?

**Comment by wonder-sk** [comment]
Just curious - do we know what kind of exception does MSVC throw at us?

**Comment by nyalldawson** [comment]
@wonder-sk 
> Just curious - do we know what kind of exception does MSVC throw at us?

I can't determine it - the Windows debugging tools are too :hankey:

**Pull request merged**